File: packages/validations/src/coldmail.ts
import { z } from "zod";

const optionalTrimmedString = (maxLength: number) =>
  z
    .string()
    .trim()
    .max(maxLength)
    .optional()
    .transform((value) => value ?? "");

const companyWebsiteUrlSchema = z.preprocess((value) => {
  if (typeof value !== "string") {
    return value;
  }

  const trimmed = value.trim();

  if (!trimmed) {
    return trimmed;
  }

  return /^https?:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`;
}, z.string().trim().url("Enter a valid company website URL.").max(500));

export const coldEmailToneSchema = z.enum([
  "warm",
  "confident",
  "direct",
  "friendly",
]);

export const coldEmailLengthSchema = z.enum([
  "concise",
  "standard",
  "detailed",
]);

export const coldEmailCallToActionSchema = z.enum([
  "conversation",
  "referral",
  "interview",
  "feedback",
]);

export const coldEmailGenerateSchema = z.object({
  resumeId: z.string().uuid(),
  jobTitle: optionalTrimmedString(200),
  companyName: optionalTrimmedString(200),
  companyWebsiteUrl: companyWebsiteUrlSchema,
  recipientName: optionalTrimmedString(120),
  recipientRole: optionalTrimmedString(160),
  jobDescription: z.string().trim().min(20).max(100_000),
  personalNote: optionalTrimmedString(1_000),
  tone: coldEmailToneSchema.default("warm"),
  length: coldEmailLengthSchema.default("standard"),
  callToAction: coldEmailCallToActionSchema.default("conversation"),
});

export const coldEmailResponseSchema = z.object({
  coldEmailId: z.string().uuid().optional(),
  resumeId: z.string().nullable().optional(),
  subject: z.string().trim().min(1).max(160),
  previewText: z.string().trim().min(1).max(240),
  body: z.string().trim().min(20).max(8_000),
  followUpSubject: z.string().trim().min(1).max(160),
  followUpBody: z.string().trim().min(20).max(8_000),
  personalizationNotes: z.array(z.string().trim().min(1).max(300)).max(6),
  qualityScore: z.number().int().min(0).max(100),
  estimatedReadTimeSeconds: z.number().int().min(10).max(300),
  companyContext: z.string().optional(),
});

export type ColdEmailGenerateInput = z.infer<typeof coldEmailGenerateSchema>;
export type ColdEmailResponse = z.infer<typeof coldEmailResponseSchema>;
export type ColdEmailTone = z.infer<typeof coldEmailToneSchema>;
export type ColdEmailLength = z.infer<typeof coldEmailLengthSchema>;
export type ColdEmailCallToAction = z.infer<
  typeof coldEmailCallToActionSchema
>;

File: apps/ai-service/app/schemas/coldmail.py
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from .resume import StructuredResume


ColdEmailTone = Literal["warm", "confident", "direct", "friendly"]
ColdEmailLength = Literal["concise", "standard", "detailed"]
ColdEmailCallToAction = Literal["conversation", "referral", "interview", "feedback"]


class ColdEmailGenerateRequest(BaseModel):
    resumeId: str | None = None
    fileUrl: str | None = None
    fileName: str | None = None
    resumeText: str | None = Field(default=None, max_length=100_000)
    structuredResume: StructuredResume | None = None
    jobTitle: str = Field(default="", max_length=200)
    companyName: str = Field(default="", max_length=200)
    companyWebsiteUrl: str = Field(default="", max_length=500)
    recipientName: str = Field(default="", max_length=120)
    recipientRole: str = Field(default="", max_length=160)
    jobDescription: str = Field(min_length=20, max_length=100_000)
    companyContext: str = Field(default="", max_length=2_000)
    personalNote: str = Field(default="", max_length=1_000)
    tone: ColdEmailTone = "warm"
    length: ColdEmailLength = "standard"
    callToAction: ColdEmailCallToAction = "conversation"

    @model_validator(mode="after")
    def clean_and_require_resume_source(self):
        has_resume_text = bool(self.resumeText and self.resumeText.strip())
        has_structured_resume = self.structuredResume is not None
        has_file = bool(self.fileUrl)

        if not (has_resume_text or has_structured_resume or has_file):
            raise ValueError("Provide resumeText, structuredResume, or fileUrl.")

        self.jobTitle = self.jobTitle.strip()
        self.companyName = self.companyName.strip()
        self.companyWebsiteUrl = self.companyWebsiteUrl.strip()
        self.recipientName = self.recipientName.strip()
        self.recipientRole = self.recipientRole.strip()
        self.jobDescription = self.jobDescription.strip()
        self.companyContext = self.companyContext.strip()
        self.personalNote = self.personalNote.strip()

        return self


class ColdEmailResponse(BaseModel):
    resumeId: str | None = None
    subject: str = Field(min_length=1, max_length=160)
    previewText: str = Field(min_length=1, max_length=240)
    body: str = Field(min_length=20, max_length=8_000)
    followUpSubject: str = Field(min_length=1, max_length=160)
    followUpBody: str = Field(min_length=20, max_length=8_000)
    personalizationNotes: list[str] = Field(default_factory=list, max_length=6)
    qualityScore: int = Field(ge=0, le=100)
    estimatedReadTimeSeconds: int = Field(ge=10, le=300)
    companyContext: str | None = Field(default=None, max_length=2_000)
